File: apps/web/src/lib/geometry.ts
/**
 * Generates a plausible farm boundary polygon from a center point and an
 * area in hectares, deterministically per farm code. This mirrors the
 * logic in backend/app/core/geometry.py but is not required to match it
 * exactly, both only stand in for real cadastral boundary data. This
 * version is used when the frontend has no backend connection and is
 * rendering the bundled demo dataset, which does not carry a boundary.
 */

const METERS_PER_DEGREE_LATITUDE = 111_320;

function seededValues(seed: string, count: number): number[] {
  let hash = 0;
  for (let i = 0; i < seed.length; i++) {
    hash = (hash << 5) - hash + seed.charCodeAt(i);
    hash |= 0;
  }
  const values: number[] = [];
  let cursor = Math.abs(hash);
  for (let i = 0; i < count; i++) {
    cursor = (cursor * 9301 + 49297) % 233280;
    const normalized = cursor / 233280;
    values.push(0.7 + normalized * 0.6);
  }
  return values;
}

export function generateFarmPolygon(
  farmCode: string,
  lat: number,
  lng: number,
  areaHectares: number
): [number, number][] {
  const areaM2 = Math.max(areaHectares, 0.1) * 10_000;
  const halfSideM = Math.sqrt(areaM2) / 2;

  const metersPerDegreeLng = METERS_PER_DEGREE_LATITUDE * Math.cos((lat * Math.PI) / 180);
  const halfLatDeg = halfSideM / METERS_PER_DEGREE_LATITUDE;
  const halfLngDeg = halfSideM / metersPerDegreeLng;

  const jitter = seededValues(farmCode, 8);

  const corners: [number, number][] = [
    [lat + halfLatDeg * jitter[0], lng - halfLngDeg * jitter[1]],
    [lat + halfLatDeg * jitter[2], lng + halfLngDeg * jitter[3]],
    [lat - halfLatDeg * jitter[4], lng + halfLngDeg * jitter[5]],
    [lat - halfLatDeg * jitter[6], lng - halfLngDeg * jitter[7]],
  ];

  return corners;
}

File: backend/app/core/geometry.py
"""
Generates a plausible farm boundary polygon from a center point and an
area in hectares. There is no real cadastral or remote sensing derived
boundary data yet, so this stands in for it. The shape is deterministic
per farm code (not randomized on every request) and roughly sized to
match the stated area, but it should not be treated as a survey accurate
boundary. Real farm polygons imported from an actual dataset, per
section 49 of the project specification, should replace this once that
data exists.
"""

import hashlib
import math

METERS_PER_DEGREE_LATITUDE = 111_320


def _seeded_values(seed: str, count: int) -> list[float]:
    digest = hashlib.md5(seed.encode("utf-8")).digest()
    return [0.7 + (digest[i % len(digest)] / 255) * 0.6 for i in range(count)]


def generate_farm_boundary(farm_code: str, lat: float, lng: float, area_hectares: float) -> list[list[float]]:
    area_m2 = max(area_hectares, 0.1) * 10_000
    half_side_m = math.sqrt(area_m2) / 2

    meters_per_degree_lng = METERS_PER_DEGREE_LATITUDE * math.cos(math.radians(lat))
    half_lat_deg = half_side_m / METERS_PER_DEGREE_LATITUDE
    half_lng_deg = half_side_m / meters_per_degree_lng

    jitter = _seeded_values(farm_code, 8)

    corners = [
        (lat + half_lat_deg * jitter[0], lng - half_lng_deg * jitter[1]),  # northwest
        (lat + half_lat_deg * jitter[2], lng + half_lng_deg * jitter[3]),  # northeast
        (lat - half_lat_deg * jitter[4], lng + half_lng_deg * jitter[5]),  # southeast
        (lat - half_lat_deg * jitter[6], lng - half_lng_deg * jitter[7]),  # southwest
    ]

    return [[round(c[0], 6), round(c[1], 6)] for c in corners]
